FEAT-#4664: Finalize compatibility support for Python 3.6

vnlitvinov opened the pull request:
<!--
Thank you for your contribution!
Please review the contributing docs: https://modin.readthedocs.io/en/latest/development/contributing.html
if you have questions about contributing.
-->

## What do these changes do?
Run almost all tests from `modin/pandas/test` with only a few cases skipped.

<!-- Please give a short brief about these changes. -->

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` <!-- you can amend your commit with a signature via `git commit -amend -s` -->
- [x] Resolves #4664
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date <!-- if you have added, renamed or removed files or directories please update the documentation accordingly -->
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release <!-- e.g. DOCS-#4077: Add release notes template to docs folder (#4078) -->

mvashishtha reviewed (commented):
@vnlitvinov I have some minor comments.

vnlitvinov commented on modin/_compat/pandas_api/latest/series.py:
```diff
@@ -17,6 +17,7 @@
 
 import numpy as np
 import pandas
+import pandas.core.groupby.generic
```
no, that's a leftover from one of my changes, removed

vnlitvinov commented on modin/_compat/pandas_api/py36/dataframe.py:
```diff
@@ -115,6 +116,35 @@ def prod(
+        tolerance=None,
+    ):
+        axis = self._get_axis_number(axis)
+        if axis == 0 and labels is not None:
+            index = labels
+        elif labels is not None:
+            columns = labels
```
indeed; I've extracted the common code to base class, thanks!

vnlitvinov commented on modin/pandas/base.py:
```diff
@@ -2284,15 +2285,7 @@ def rmod(
-        axis,
-        closed,
-        method,
+        self, window, min_periods, center, win_type, *args, **kw
```
I guess this is how I usually specify things; search in Modin codebase shows that usually it's `kwargs`, so I'll fix that

vnlitvinov commented on modin/pandas/test/dataframe/test_default.py:
```diff
@@ -74,8 +76,20 @@
-        ("flags", None),
-        ("set_flags", lambda df: {"allows_duplicate_labels": False}),
+        pytest.param(
+            *("flags", None),
```
yes 😃 this is to trick `black` to _not_ split the two in two lines 🙃 IMHO in one line it's clearer to see that those two are closely related

vnlitvinov commented on modin/pandas/test/dataframe/test_default.py:
```diff
@@ -74,8 +76,20 @@
+            ),
+        ),
+        pytest.param(
+            *("set_flags", lambda df: {"allows_duplicate_labels": False}),
```
same, this is to trick `black`

vnlitvinov commented on modin/pandas/test/dataframe/test_udf.py:
```diff
@@ -79,6 +80,7 @@ def test_agg_apply(axis, func, op):
     eval_general(
         *create_test_dfs(test_data["float_nan_data"]),
         lambda df: getattr(df, op)(func, axis),
+        check_exception_type=PandasCompatVersion.CURRENT != PandasCompatVersion.PY36,
```
added

vnlitvinov commented:
Thanks @mvashishtha, I think I've addressed all your comments, please take another look!

@devin-petersohn @RehanSD 
For reference, here's the set of `groupby`-related tests which are skipped now because older pandas gives weird exceptions from within our implementation. I decided to skip them instead of trying to fix, because they're mostly edge cases related to zero-rows dataframes IIRC.
To see those failures one has to remove skippings in the `test_groupby.py`.
```
========================================================================= short test summary info =======================================================================
FAILED modin/pandas/test/test_groupby.py::test_groupby_with_kwarg_dropna[col_name_multi_by-True] - IndexError: index -1 is out of bounds for axis 0 with size 0
FAILED modin/pandas/test/test_groupby.py::test_groupby_with_kwarg_dropna[col_name+level_name-True] - IndexError: index -1 is out of bounds for axis 0 with size 0
FAILED modin/pandas/test/test_groupby.py::test_groupby_with_kwarg_dropna[col_name-False] - TypeError: 'NoneType' object is not subscriptable
FAILED modin/pandas/test/test_groupby.py::test_groupby_with_kwarg_dropna[col_name-True] - TypeError: Expected tuple, got str
=============================================================== 4 failed, 12 passed, 85 warnings in 54.51s ==============================================================
```

mvashishtha reviewed (dismissed):
LGTM @vnlitvinov, thanks for addressing comments!

YarShev commented on modin/_compat/pandas_api/py36/base.py:
```diff
@@ -29,8 +29,8 @@ class Python36CompatibleBasePandasDataset(BaseCompatibleBasePandasDataset):
     """Compatibility layer for 'Python 3.6 pandas' for Dataset."""
 
     def _validate_ascending(self, ascending):  # noqa: PR01, RT01
```
Do we need this method?

vnlitvinov replied:
this is for uniform API between different compatibility classes, we'll likely add more in the future...

YarShev replied:
I see, thanks!

YarShev commented on modin/pandas/base.py:
```diff
@@ -2284,15 +2285,7 @@ def rmod(
-        axis,
-        closed,
-        method,
+        self, window, min_periods, center, win_type, *args, **kw
```
```suggestion
        self, window, min_periods, center, win_type, *args, **kwargs
```

YarShev commented on modin/pandas/base.py:
```diff
@@ -2306,10 +2299,8 @@ def _rolling(
-                closed=closed,
-                method=method,
+                *args,
+                **kw,
```
```suggestion
                **kwargs,
```

YarShev commented on modin/pandas/base.py:
```diff
@@ -2319,10 +2310,8 @@ def _rolling(
-            closed=closed,
-            method=method,
+            *args,
+            **kw,
```
```suggestion
            **kwargs,
```

YarShev commented on modin/pandas/test/dataframe/test_udf.py:
```diff
@@ -399,4 +406,7 @@ def test_transform(data, func):
     eval_general(
         *create_test_dfs(data),
         lambda df: df.transform(func),
+        # older pandas gives different exception type compared to Modin and current pandas,
```
I wonder if the user expect the exception from Modin, which was in older pandas versions, should we raise that exception?

vnlitvinov replied:
this would complicate our base without much profit; I'd rather not try to support every nook and cranny of older pandas and wait until users explicitly request that support

YarShev replied:
makes sense

YarShev commented on modin/pandas/test/test_groupby.py:
```diff
@@ -1305,11 +1375,26 @@ def test_groupby_multiindex(groupby_kwargs):
             id="col_name+level_name",
+            marks=pytest.mark.skipif(
+                PandasCompatVersion.CURRENT == PandasCompatVersion.PY36,
+                reason="Weird exception in Modin internals",
```
Don't we want to file an issue for this?

vnlitvinov replied:
Generally, we would, but I skipped the tests which were not working when a groupby resulted in zero-row dataframe, and I think this is were the line should be drawn.

I mean, I don't see a point of submitting an issue which no one is ever planning to fix.

YarShev replied:
ok, probably the message above makes sense for this too

> wait until users explicitly request that support

YarShev commented on modin/pandas/test/test_groupby.py:
```diff
@@ -1305,11 +1375,26 @@ def test_groupby_multiindex(groupby_kwargs):
+            id="col_name",
+            marks=pytest.mark.skipif(
+                PandasCompatVersion.CURRENT == PandasCompatVersion.PY36,
+                reason="Weird exception in Modin internals",
```
same

vnlitvinov replied:
https://github.com/modin-project/modin/pull/4800#discussion_r944652994

YarShev commented on modin/pandas/test/test_groupby.py:
```diff
@@ -1305,11 +1375,26 @@ def test_groupby_multiindex(groupby_kwargs):
             id="col_name_multi_by",
+            marks=pytest.mark.skipif(
+                PandasCompatVersion.CURRENT == PandasCompatVersion.PY36,
+                reason="Weird exception in Modin internals",
```
same

vnlitvinov replied:
https://github.com/modin-project/modin/pull/4800#discussion_r944652994

YarShev commented on modin/pandas/test/test_rolling.py:
```diff
@@ -91,7 +92,10 @@ def test_dataframe(data, window, min_periods, win_type, axis):
         df_equals(modin_rolled.kurt(), pandas_rolled.kurt())
-        df_equals(modin_rolled.apply(np.sum), pandas_rolled.apply(np.sum))
+        if PandasCompatVersion.CURRENT != PandasCompatVersion.PY36:
+            # skip this on pandas 1.1 as pandas raise an exception in there
```
Don't we want to file an issue for this?

vnlitvinov replied:
https://github.com/modin-project/modin/pull/4800#discussion_r944652994

YarShev reviewed (commented):
The changes are pretty good! Left some comments.
